### ISSUE TITLE
RevEng: Prefer [StringLength] over [MaxLength] on string properties

### ISSUE THREAD

**Issue by bricelam** [aspnet/EntityFramework#8468]
I've found that `[StringLength]` is generally more supported than `[MaxLength]`. It would be nice if `Scaffold-DbContext -DataAnnotations` used this instead on `string` properties.

**Comment by ajcvickers** [comment]
Putting in 2.0 because this should be very quick fix.

**Comment by lajones** [comment]
Note: we also currently generate `MaxLength` for `varbinary` columns. `StringLength` might seem odd on those.

**Comment by lajones** [comment]
Also found this comment:

```
        // nvarchar is the default column type for string properties,
        // so we don't need to define it using HasColumnType() and removing
        // the column type allows the HasMaxLength() API to have effect.
        // Similarly for varbinary and byte[] properties.
```

Not sure if HasColumnType() and StringLength have the same interaction?

**Comment by bricelam** [comment]
Correct, just on `string` properties, not `byte[]`.

Re interaction with `HasColumnType()`: The interaction will be the same as it was with `MaxLength`. They're handled by nearly identical conventions that run back-to-back.

**Comment by lajones** [comment]
@bricelam So, just to be clear: for fluent API scaffolding you want the `HasMaxLength()` fluent API regardless of property type (as there is no `PropertyBuilder.HasStringLength()` API); for scaffolding with `-DataAnnotations`  you want `byte[]` properties to still have the `[MaxLength]` attribute but scaffolded `string` properties to use the `[StringLength]` attribute instead. Is that correct?

(Note to self: this is handled in [ModelConfiguration.AddMaxLengthConfiguration()](https://github.com/aspnet/EntityFramework/blob/dev/src/EFCore.Design/Scaffolding/Configuration/Internal/ModelConfiguration.cs#L500))

**Comment by bricelam** [comment]
Yep. Sounds right

**Comment by ajcvickers** [comment]
@lajones Can you elaborate on why this is flagged as a breaking change?

**Comment by lajones** [comment]
@ajcvickers Only breaking in a minor way - it will produce different code than people were used to before. If they (or any software into which the scaffolded classes get fed) were expecting/doing anything with that attribute themselves then we will break them.

**Comment by ajcvickers** [comment]
@lajones I can see your point, but this isn't something that we usually consider a breaking change.

**Comment by smitpatel** [comment]
Taking this one over as it is just small change in my present work in refactoring how code files are written.

**Comment by smitpatel** [comment]
Fixed in #8583 
commit 05249433991ce11ab8df15a66354f82bb300e341
